### PR TITLE
Layers support

### PR DESCRIPTION
**Pull request by ntunin**
## Description
There was implemented an ability to configure layer frame via framezilla
Please notice that maker was divided with different  extensions due to large file and body length

## Related tasks
[ISD-162](https://rosberry.atlassian.net/browse/ISD-162)

## TODOs
Write tests

### CODE REVIEW

File: Sources/Extensions/CALayer+Relations.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    /// CenterY relation of current view.
+
+    var nui_centerY: RelationView<VerticalRelation> {
+        return RelationView<VerticalRelation>(view: .layer(self), relation: .centerY)
+    }
+
+    /// Safe area
+
+    @available(iOS 11.0, *)
+    var nui_safeArea: SafeAreaRelationCollection {
```
Comment (disabled) [inline Sources/Extensions/CALayer+Relations.swift]: I don't think we should have `nui_safeArea` on layers at all.

File: Sources/Parameters.swift [hunk 2]
```diff
@@ -18,11 +18,11 @@ final class ValueParameter {
 
 final class SideParameter {
 
-    unowned let view: UIView
```
Comment (disabled) [inline Sources/Parameters.swift:21]: Why did you remove `unowned`? Same for other places like this in `Relations.swift`.

File: Sources/ViewType.swift [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                return layer.convert(rect, from: nil)
+            }
+        }
+        switch (self, view) {
+        case (.view(let superview), .view(let view)):
+            return superview.convert(rect, from: view)
+        case (.layer(let superlayer), .layer(let layer)):
+            return superlayer.convert(rect, from: layer)
+        default:
+            return rect
```
Comment (disabled) [inline Sources/ViewType.swift]: You are missing a couple of cases here - it won't work if you'll add relation between a layer and a view when rect conversion is required.

**Review by disabled — CHANGES_REQUESTED** [review]
Overall, I don't like this approach too much. There's a ton of switches in ViewType which basically do the same thing on one hand, and on other, you have `sizeToFit` & `sizeThatFits` maker methods that won't do anything for layers. 

I would get rid of `ViewType` abstraction altogether, reworking `Maker` instead, splitting it into 2 makers, actually. First (the base) one will work with layers and only have funcs that make sense for layers; the second one will be a subclass, which will add funcs that make sense for views (like `sizeToFit`).

What others think on this one?

File: Sources/Extensions/CALayer+Installer.swift [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        }
+        set {
+            objc_setAssociatedObject(self, &nxStateTypeAssociationKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
+    }
+}
+
+public extension CALayer {
+
+    /// Configures frame of current view for special state.
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
    /// Configures frame of current layer for special state.
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+import UIKit
+
+public extension CALayer {
+
+    /// Apply new configuration state without frame updating.
+    ///
+    /// - note: Use `DEFAULT_STATE` for setting the state to the default value.
+
+    var nx_state: AnyHashable {
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]: Actually, we don't use states, so maybe we should deprecate it.

File: Sources/Extensions/CALayer+Installer.swift [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    func configureFrame(state: AnyHashable = DEFAULT_STATE, installerBlock: InstallerBlock) {
+        guard self.superlayer != nil else {
+            return
+        }
+
+        Maker.configure(view: .layer(self), for: state, installerBlock: installerBlock)
+    }
+
+    /// Configures frame of current view for special states.
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
    /// Configures frame of current layer for special states.
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    ///
+    /// - note: Don't forget about `DEFAULT_VALUE`.
+    ///
+    /// - parameter states:         The states for which you configure frame.
+    /// - parameter installerBlock: The installer block within which you can configure frame relations.
+
+    func configureFrame(states: [AnyHashable], installerBlock: InstallerBlock) {
+        guard self.superlayer != nil else {
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
        guard superlayer != nil else {
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    ///
+    /// - parameter state:          The state for which you configure frame. Default value: `DEFAULT_STATE`.
+    /// - parameter installerBlock: The installer block within which you can configure frame relations.
+
+    func configureFrame(state: AnyHashable = DEFAULT_STATE, installerBlock: InstallerBlock) {
+        guard self.superlayer != nil else {
+            return
+        }
+
+        Maker.configure(view: .layer(self), for: state, installerBlock: installerBlock)
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
        configureFrame(states: [state], installerBlock: installerBlock)
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+    /// Configures frames of the views for special state.
+    ///
+    /// - note: When you configure frame without implicit state parameter (default value), frame configures for the `DEFAULT_STATE`.
+    ///
+    /// - parameter state:          The state for which you configure frame. Default value: `DEFAULT_STATE`.
+    /// - parameter installerBlock: The installer block within which you can configure frame relations.
+
+    func configureFrames(state: AnyHashable = DEFAULT_STATE, installerBlock: InstallerBlock) {
+        for view in self {
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
        for layer in self {
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    /// Configures frames of the views for special states.
+    ///
+    /// - note: Don't forget about `DEFAULT_VALUE`.
+    ///
+    /// - parameter states:         The states for which you configure frames.
+    /// - parameter installerBlock: The installer block within which you can configure frame relations.
+
+    func configureFrames(states: [AnyHashable], installerBlock: InstallerBlock) {
+        for view in self {
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
        for layer in self {
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            view.configureFrame(states: states, installerBlock: installerBlock)
+        }
+    }
+}
+
+public extension Collection where Iterator.Element: CALayer {
+
+    /// Configures all subview within a passed container.
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
    /// Configures all sublayers within a passed container.
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    /// - note: If you don't use a static width for instance, important to understand, that it's not correct to call 'left' and 'right' relations together by subviews,
+    ///         because `container` sets width relatively width of subviews and here is some ambiguous.
+    ///
+    /// - parameter container:                The layer where a container will be added.
+    /// - parameter relation:            The relation of `ContainerRelation` type.
+    ///     - `width`:                   The width of a container. If you specify a width only a dynamic height will be calculated.
+    ///     - `height`:                  The height of a container. If you specify a height only a dynamic width will be calculated.
+    ///     - `horizontal(left, right)`: The left and right insets of a container. If you specify these parameters only a dynamic height will be calculated.
+    ///     - `vertical(top, bottom)`:   The top and bototm insets of a container. If you specify these parameters only a dynamic width will be calculated.
+    /// - parameter installerBlock:      The installer block within which you should configure frames for all subviews.
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]: Please update docs with layers

File: Sources/Extensions/CALayer+Installer.swift [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    /// - note: It automatically adds all subviews to the container. Don't add subviews manually.
+    /// - note: If you don't use a static width for instance, important to understand, that it's not correct to call 'left' and 'right' relations together by subviews,
+    ///         because `container` sets width relatively width of subviews and here is some ambiguous.
+    ///
+    /// - parameter container:                The layer where a container will be added.
+    /// - parameter relation:            The relation of `ContainerRelation` type.
+    ///     - `width`:                   The width of a container. If you specify a width only a dynamic height will be calculated.
+    ///     - `height`:                  The height of a container. If you specify a height only a dynamic width will be calculated.
+    ///     - `horizontal(left, right)`: The left and right insets of a container. If you specify these parameters only a dynamic height will be calculated.
+    ///     - `vertical(top, bottom)`:   The top and bototm insets of a container. If you specify these parameters only a dynamic width will be calculated.
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
    ///     - `vertical(top, bottom)`:   The top and bottom insets of a container. If you specify these parameters only a dynamic width will be calculated.
```

File: Sources/Extensions/CALayer+Installer.swift [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    /// - parameter installerBlock:      The installer block within which you should configure frames for all subviews.
+    ///
+    /// - returns: Container view.
+
+    func container(in layer: CALayer, relation: ContainerRelation? = nil, installerBlock: () -> Void) -> CALayer {
+        let container: CALayer
+        if let superLayer = self.first?.superlayer {
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]:
```suggestion
        if let superLayer = first?.superlayer {
```

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

**Comment by artemnovichkov** [comment]
Agreed with @disabled, and having two makers is a good idea. Anyway, thanks for PR, I'm waiting to use it so long.

File: Example/FramezillaExample/ViewController.swift [hunk 15]
```diff
@@ -22,6 +22,10 @@ class ViewController: UIViewController {
     let label2 = UILabel()
     let label3 = UILabel()
 
+    let layer1 = CALayer()
+    let layer2 = CALayer()
+    let layer3 = CALayer()
```
Comment (Antowkos) [inline Example/FramezillaExample/ViewController.swift]: We can configure these layers (and all other subviews) as `lazy var`. It will allow us to reduce `viewDidLoad`

File: Sources/Extensions/UIView+Installer.swift [hunk 16]
```diff
@@ -11,8 +11,8 @@ import ObjectiveC
 
 public let DEFAULT_STATE = "DEFAULT STATE"
 
-private var stateTypeAssociationKey: UInt8 = 0
-private var nxStateTypeAssociationKey: UInt8 = 1
+var stateTypeAssociationKey: UInt8 = 0
+var nxStateTypeAssociationKey: UInt8 = 1
```
Comment (Antowkos) [inline Sources/Extensions/UIView+Installer.swift:15]: I suppose they should remain private

**Review by Antowkos — CHANGES_REQUESTED** [review]
I agree with @disabled. Also I have a question about `mask`s for layer. Will they be properly laid out? I think we should also add `mask`s in example project

File: Sources/Parameters.swift (hunk 2 above)
Comment (ntunin) [reply Sources/Parameters.swift:21]: because this `view` replaced by view type that is not class

File: Sources/Extensions/CALayer+Installer.swift (hunk 5 above)
Comment (ntunin) [reply Sources/Extensions/CALayer+Installer.swift]: What is the reason of deprecation for message?

File: Sources/Extensions/UIView+Installer.swift (hunk 16 above)
Comment (ntunin) [reply Sources/Extensions/UIView+Installer.swift:15]: they are still module private an used for both extensions `UIView+Installer` and `CALayer_Installer`

**Review by stason314 — APPROVED** [review]

**Comment by ntunin** [comment]
I have improved `ViewType` with `strategy` pattern to avoid extra switches. Also i've placed common maker logic for any ViewType to private methods. View and layer frame configurations now produce makers of different types that use this common logic to avoid code duplication

**Review by ReshetovE — APPROVED** [review]

File: Sources/Extensions/CALayer+Installer.swift (hunk 5 above)
Comment (artemnovichkov) [reply Sources/Extensions/CALayer+Installer.swift]: I'm not sure what you mean... Ping me in Slack DMs

File: Sources/Maker/Maker+Edges.swift [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+            case .right:
+                _right(inset: insets.right)
+            case .top:
+                _top(inset: insets.top)
+            default:
+                return
+            }
+        }
+    }
+}
```
Comment (artemnovichkov) [inline Sources/Maker/Maker+Edges.swift]: I don't like the concept with `_` functions, because you still should copy implementations. What's about using protocols with default implementations?
```suggestion
public protocol Edges {

    @discardableResult func edges(top: Number?, left: Number?, bottom: Number?, right: Number?) -> Self
}

extension Maker: Edges {

    public func edges(top: Number? = nil, left: Number? = nil, bottom: Number? = nil, right: Number? = nil) -> Self {
        apply(self._top, top)
        apply(self._left, left)
        apply(self._bottom, bottom)
        apply(self._right, right)
        return self
    }
}
```

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: Sources/Extensions/CALayer+Installer.swift [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+    /// - note: Don't forget about `DEFAULT_VALUE`.
+    ///
+    /// - parameter states:         The states for which you configure frames.
+    /// - parameter installerBlock: The installer block within which you can configure frame relations.
+
+    func configureFrames(states: [AnyHashable], installerBlock: InstallerBlock) {
+        for layer in self {
+            layer.configureFrame(installerBlock: installerBlock)
+        }
+    }
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift]: You didn't use state inside the function

File: Sources/Extensions/CALayer+Installer.swift [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+
+        if let width = relationWidth {
+            container.frame.size.width = width
+        }
+
+        if let height = relationHeight {
+            container.frame.size.height = height
+        }
+
+        installerBlock()
```
Comment (artemnovichkov) [inline Sources/Extensions/CALayer+Installer.swift:106]: Why installerBlack is called twice?

File: Sources/Extensions/Maker+Configurations.swift [hunk 20]
```diff
@@ -20,26 +20,38 @@ postfix operator >>
...
 /// End frame configuration.
 
 public postfix func >> (maker: Maker) {
-    if (maker.view.nx_state as? String) == DEFAULT_STATE {
-        maker.configureFrame()
-    }
+    maker.configureFrame()
 }
 
+public typealias UIViewInstallerBlock = (UIViewMaker) -> Void
```
Comment (artemnovichkov) [inline Sources/Extensions/Maker+Configurations.swift]: Is it possible to remove `UIViewInstallerBlock`?

File: Sources/Maker/Maker+Edges.swift [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+            switch side {
+            case .bottom:
+                bottom(inset: insets.bottom)
+            case .left:
+                left(inset: insets.left)
+            case .right:
+                right(inset: insets.right)
+            case .top:
+                top(inset: insets.top)
+            default:
```
Comment (artemnovichkov) [inline Sources/Maker/Maker+Edges.swift:63]: I know it's not related to the PR, but what's about other sides?
```swift
case .horizontal:
    left(inset: insets.left)
    right(inset: insets.right)
```

File: Sources/Maker/UIViewMaker.swift [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+import UIKit
+
+public final class UIViewMaker: Maker {
+
+    /// Resizes the current view so it just encloses its subviews.
+    ///
+    /// - returns: `Maker` instance for chaining relations.
+
+    unowned let uiView: UIView
```
Comment (artemnovichkov) [inline Sources/Maker/UIViewMaker.swift]:
```suggestion
    unowned let view: UIView
```

File: Sources/ViewType.swift [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    var safeAreaInsets: UIEdgeInsets {
+        if #available(iOS 11.0, *) {
+            return view.safeAreaInsets
+        }
+        return .zero
+    }
+
+    init(_ view: UIView) {
```
Comment (artemnovichkov) [inline Sources/ViewType.swift]:
```suggestion
    init(view: UIView) {
```

File: Sources/ViewType.swift [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+        }
+        switch view {
+        case let .view(view):
+            return self.layer.convert(rect, from: view.layer)
+        case let .layer(layer):
+            return self.layer.convert(rect, from: layer)
+        }
+    }
+
+    init(_ layer: CALayer) {
```
Comment (artemnovichkov) [inline Sources/ViewType.swift]:
```suggestion
    init(layer: CALayer) {
```

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

**Comment by artemnovichkov** [comment]
BTW, some tests won't passed after last changes.

File: Sources/Extensions/Maker+Configurations.swift (hunk 20 above)
Comment (ntunin) [reply Sources/Extensions/Maker+Configurations.swift]: can not find the reasonable purpose of removing

File: Sources/Maker/UIViewMaker.swift (hunk 22 above)
Comment (ntunin) [reply Sources/Maker/UIViewMaker.swift]: No, `Maker` already has `view: ViewType`

File: Sources/Extensions/CALayer+Installer.swift (hunk 19 above)
Comment (ntunin) [reply Sources/Extensions/CALayer+Installer.swift:106]: Without it container and its content appears properly only on the next layout

Comment (ntunin) [reply Sources/Extensions/CALayer+Installer.swift:106]: legacy code, nested from UIView+Installer

File: Sources/Extensions/HasRelations.swift [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    /// Height relation of current instance.
+
+    public var nui_height: RelationView<SizeRelation> {
+        return RelationView<SizeRelation>(view: viewType, relation: .height)
+    }
+
+    /// Left relation of current instance.
+
+    public var nui_left: RelationView<HorizontalRelation> {
+        return RelationView<HorizontalRelation>(view: viewType, relation: .left)
```
Comment (zhvrnkov) [inline Sources/Extensions/HasRelations.swift]: You can omit `<HorizontalRelation>` and return here and in all other places

**Review by Antowkos — APPROVED** [review]

File: Sources/ViewType.swift [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+            switch self {
+            case .view(let view):
+                layout = UIViewLayout(view: view)
+            case .layer(let layer):
+                layout = CALayerLayout(layer: layer)
+            }
+            objc_setAssociatedObject(self, &layoutAssociationKey, layout, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+            return layout
+        }
+    }
```
Comment (disabled) [inline Sources/ViewType.swift]: That's why you shouldn't use enums when you need to have a stored object.

File: Sources/ViewType.swift [hunk 27]
```diff
@@ -0,0 +1,225 @@
...
+        for subview in subviews {
+            if subview.layout.contains(view) {
+                return true
+            }
+        }
+        return false
+    }
+}
+
+final class UIViewLayout: Layout {
```
Comment (disabled) [inline Sources/ViewType.swift]: Naming is incorrect, this class is not Obj-C and it doesn't belong to UIKit namespace.

File: Sources/ViewType.swift [hunk 28]
```diff
@@ -0,0 +1,225 @@
...
+        }
+        return .zero
+    }
+
+    init(view: UIView) {
+        self.view = view
+    }
+}
+
+final class CALayerLayout: Layout {
```
Comment (disabled) [inline Sources/ViewType.swift]: Naming is incorrect, this class is not Obj-C and it doesn't belong to CoreAnimation namespace.

File: Sources/ViewType.swift [hunk 29]
```diff
@@ -0,0 +1,225 @@
...
+            view.bounds
+        }
+        set {
+            view.bounds = newValue
+        }
+    }
+
+    var cornerRadius: CGFloat {
+        get {
+            view.layer.cornerRadius
```
Comment (disabled) [inline Sources/ViewType.swift]: Why do you insist on having duplicate logic for views and frames? This layout should be a subclass of LayerLayout.

File: Sources/ViewType.swift [hunk 30]
```diff
@@ -0,0 +1,225 @@
...
+            return false
+        }
+    }
+}
+
+protocol Layout: class {
+    var superview: ViewType? { get }
+    var subviews: [ViewType] { get }
+    var frame: CGRect { get set}
+    var bounds: CGRect { get set}
```
Comment (disabled) [inline Sources/ViewType.swift]:
```suggestion
    var frame: CGRect { get set }
    var bounds: CGRect { get set }
```

File: Sources/Maker/UIViewMaker.swift [hunk 31]
```diff
@@ -0,0 +1,235 @@
+//
+//  Copyright © 2020 Rosberry. All rights reserved.
+//
+
+import UIKit
+
+public final class UIViewMaker: Maker {
```
Comment (disabled) [inline Sources/Maker/UIViewMaker.swift]: Naming is incorrect, this class is not Obj-C and it doesn't belong to UIKit namespace.

File: Sources/Maker/UIViewMaker.swift [hunk 32]
```diff
@@ -0,0 +1,235 @@
...
+    /// Resizes the current view so it just encloses its subviews.
+    ///
+    /// - returns: `Maker` instance for chaining relations.
+
+    unowned let uiView: UIView
+
+    override init(view: ViewType) {
+        guard case let .view(view) = view else {
+            fatalError("UIViewMaker was initialized with wrong ViewType")
+        }
```
Comment (disabled) [inline Sources/Maker/UIViewMaker.swift]: You can just add separate init that accepts UIView to avoid this hack.

**Review by disabled — CHANGES_REQUESTED** [review]

File: Sources/ViewType.swift (hunk 27 above)
Comment (ntunin) [reply Sources/ViewType.swift]: This naming assumes that this class is layout for UIView

File: Sources/ViewType.swift (hunk 28 above)
Comment (ntunin) [reply Sources/ViewType.swift]: This naming assumes that this class is layout for CALayer. What is the name it should have by your opinion?

File: Sources/ViewType.swift (hunk 26 above)
Comment (ntunin) [reply Sources/ViewType.swift]: Even without enum i need associated object to store some instance that provides access to layout properties

File: Sources/ViewType.swift (hunk 28 above)
Comment (disabled) [reply Sources/ViewType.swift]: Just `ViewLayout` / `LayerLayout` would do just fine. I completely understand what's intention here, it's just mixing naming schemes what's unfortunate here. Since UIView & CALayer are Obj-C classes, UI & CA are class prefixes. And we don't use class prefixes in swift code, especially class prefixes of existing frameworks.

File: Sources/ViewType.swift (hunk 26 above)
Comment (disabled) [reply Sources/ViewType.swift]: Well, no, you don't need an associated object at all. You can do something like this:
```
public class ElementType {
    static func view(_ view: UIView) -> ElementType {
        return ViewType(view: view)
    }

    static func layer(_ layer: CALayer) -> ElementType {
        return LayerType(layer: layer)
    }

    let layout: Layout
    let layer: CALayer

    init(layer: CALayer, layout: Layout) {
        self.layer = layer
        self.layout = layout
    }
}

public class ViewType: ElementType {
   let view: UIView

    init(view: UIView) {
        self.view = view
        let layout = ViewLayout(view: view)
        super.init(layer: view.layer, layout: layout)
    }
}

public class LayerType: ElementType {
    init(layer: CALayer) {
        let layout = LayerLayout(layer: layer)
        super.init(layer: layer, layout: layout)
    }
}
```
Switched naming here a bit, but you get the idea. 
You're losing convenient `switch` by getting rid of enum, but the best part is that you won't need to use `switch` at all, because you'll be able to access layers directly.

**Review by disabled — APPROVED** [review]
👍

**Review by artemnovichkov — APPROVED** [review]

**Comment by disabled** [comment]
@rosberry/ios, unless anyone has any objections or change requests, this one will become a v5.0 release on June 26th.
